### PR TITLE
RFD 114: Partial cache healthiness

### PR DESCRIPTION
**Pull request by andreiko**
Issue: https://github.com/gravitational/teleport/issues/21586

Rendered: https://github.com/gravitational/teleport/blob/andrey/rfd-partial-cache/rfd/0xxx-partial-cache-healthiness.md

### CODE REVIEW

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [inline rfd/0xxx-partial-cache-healthiness.md:135]: We could also sneak this information into inner implementations as a value on `ctx` like we do with loggers and other things. I'm not a fan of that approach because hiding things away can lead to future programming errors, but it can make this API look tidier on the outside.

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: Makes sense, updated the doc in https://github.com/gravitational/teleport/pull/23494/commits/02749743f2fd61bc4e18fa7a8a0e4bedc8117af1

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: Yeah I agree, wasn't too confident about that name. Renamed into `KindsRejectedInFlight` in https://github.com/gravitational/teleport/pull/23494/commits/5f5491c86e3762ac9d3be986d25fbbe6f6d323d1

Any thoughts on passing that implicitly via `Context`? https://github.com/gravitational/teleport/pull/23494#discussion_r1146903370

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: good idea, changed that where appropriate in https://github.com/gravitational/teleport/pull/23494/commits/c891de6b3351e4c96638fddc226e3d67ddc21dc8

File: rfd/0xxx-partial-cache-healthiness.md
Comment (fspmarshall) [inline rfd/0xxx-partial-cache-healthiness.md:172]: note: in order to prevent concurrency bugs, we'll need to do allocate a new slice for `w.Kinds` each time we remove an element.

Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:172]: That was the idea, but it's good to make a note of that https://github.com/gravitational/teleport/pull/23494/commits/9a622b5224908a33dea963271332ba45a03213f7

File: rfd/0xxx-partial-cache-healthiness.md
Comment (fspmarshall) [inline rfd/0xxx-partial-cache-healthiness.md:237]: I really like the generics proposal.  Looking it over, I do think impl might be complex enough that it'd be best to categorize this as a "nice to have", or "future work".

Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:237]: I was able to move a couple collectors to this new interface during prototyping and it wasn't too bad. The worst part was that some of the kinds don't have an exactly corresponding interface in the `services` package, e.g.:

```go
package services

type RoleGetter interface {
	GetRole(ctx context.Context, name string) (types.Role, error)
}
```

This one is missing `GetRoles()` required for `Cache`, so I had to define a private `roleGetter` interface in collections.go.

Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:237]: So you suggest to go with the simpler option from https://github.com/gravitational/teleport/issues/21586 for now?

**Review by fspmarshall — APPROVED** [review]
Proposal LGTM.

A general note: this feature will probably make it easier for people to accidentally write logic that will _always_ experience rejections (e.g. adding a watch kind to an agent's cache without adding it to auth's cache as well).  It may be worth trying to improve our tests to compensate.  Some ideas:

- Make caches configurable so that we can have them run with `AllowRejectedKinds = false` during most tests.
- Add test coverage that explicitly ensures that all the watch kinds in `For<service>` constructors are present in `ForAuth`.
- Add a "fuzz" option that randomly disables `AllowRejectedKinds` 50% of the time (maybe this is a terrible idea... not sure... sounds neat tho).

**Comment by andreiko** [comment]
Thanks for taking the time to review, Forrest! Do you have any suggestions on who else I should add to required approvers?

File: rfd/0xxx-partial-cache-healthiness.md
Comment (fspmarshall) [reply rfd/0xxx-partial-cache-healthiness.md:237]: Just a gut reaction... point is just that I wouldn't consider generics a blocker since we have a reasonable fallback in `read(kind)`.  If you feel good about it, go for it!  But if it ends up feeling like a pain, I wouldn't invest too much energy into it at this stage.

My guess is that `ListResources` would be the trickiest to handle correctly in the context of generics (`ListResources` is _always_ causing trouble lol).  Maybe take a stab at applying the proposed generic strategy to `ListResources`... if it ends up being super difficult, that might be a good indicator that we should just go with `read(kind)` for now.

**Comment by fspmarshall** [comment]
@andreiko I'd wait for approval from at least one of @rosstimothy and @espadolini.

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:237]: Got it. Thanks for the heads up about `ListResources`!

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:159]: It's true that `WatchStatus` is not a real resource, but to be assignable to `event.Resource` it still needs to implement the `Resource` interface and have methods like`GetKind()`, `GetMetadata()`, etc. We could omit the resource header in protobuf and just implement those getters as returning string literals, e.g.

```go
func (m *WatchStatusV1) GetKind() string {
	return "watch_status"
}
```

Personally, I would rather pretend that it's a real resource for consistency and in case we decide to pass more information via the headers in the future.

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:86]: Actually, even before that, if an old provider doesn't recognize some of the requested kinds, it will return a failure, but if it recognizes all of them, it doesn't matter if it's old or new.

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:177]: I like that! The `OpInit` resource could look like this then:

```protobuf
message WatchStatusSpecV1 {
  repeated WatchKind WatchedKinds = 1;
}
```

Then, a consumer would need to handle two variants of `OpInit` event:
 - resource missing: this is a response from an old Auth service
 - resource is present: only specified kinds are considered successful

@fspmarshall what do you think about this?

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md:54]: Added a note https://github.com/gravitational/teleport/pull/23494/commits/55004edaf5fdf3c726e5e1cd2f1bb5f00094317b

File: rfd/0xxx-partial-cache-healthiness.md
Comment (espadolini) [reply rfd/0xxx-partial-cache-healthiness.md:177]: While it would be mildly annoying and not really worth it (especially if we return `WatchedKinds` instead of `RejectedKinds` in `OpInit`), any implementation of `NewWatcher` that wraps and delegates to an upstream `Events` can just return a wrapper for the `types.Watcher` instead, which can alter anything it wants.

Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: I updated the doc https://github.com/gravitational/teleport/pull/23494/commits/9261e4fdc19478d6d55c54c95a499ce7fc0934bb

```protobuf
message Watch {
   repeated WatchKind Kinds = 1;
   bool AllowPartialSuccess = 2;
}

message WatchStatusSpecV1 {
  repeated WatchKind Kinds = 1;
}
```

This way we get cleaner API and don't need to update wrappers too much.

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: Good idea. The two notes were actually saying the same thing, so I rephrased and merged them together in https://github.com/gravitational/teleport/commit/197ba0a0534d984a0d886e1f7504efe0d8d713a0

**Review by rosstimothy — APPROVED** [review]

**Review by espadolini — APPROVED** [review]

File: rfd/0xxx-partial-cache-healthiness.md
Comment (andreiko) [reply rfd/0xxx-partial-cache-healthiness.md]: `ListResources` using a generic `readGuard` is shaping up to look something like this https://github.com/gravitational/teleport/commit/f93dbfb5e1d2f3157d56bd87581b93d92e546011
It's still mostly an exception because there's no corresponding collection.

I'll continue converting other collections for for now, but let me know if this seems over the top — I can cut it short and get back to the simple implementation.

**Pull request merged**
